### ISSUE TITLE
On Ubuntu 16.04 (xenial): compile errors w/ Boost

### ISSUE THREAD

**Issue by dstndstn** [davidsd/sdpb#35]
Hi,

I'm an IT staffer at Perimeter Institute (Canada) trying to install a build from source on our Ubuntu 16.04 cluster.  Following your excellent INSTALL.md -- thank you!

One thing -- not sure if this is intended or not -- it looks like you merged the elemental branch to master a while back.  The INSTALL.md instructions say to check out the elemental branch.  I encountered a different error there, but then tried the master branch and at the ./waf stage I am getting:

```
[88/94] Compiling src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx
12:06:07 runner ['mpicxx', '-Wall', '-Wextra', '-O3', '-D SDPB_VERSION_STRING="2.1.2-2-g7cadb27"', '-std=c++14', '-I/cm/shared/apps/elemental/spdb-fork/include', '-I/usr/include/libxml2', '-DHAVE_GMPXX_H=
1', '-DHAVE_MPFR_H=1', '-DHAVE_LIBXML2=1', '../src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx', '-c', '-o/home/dlang/spdb-build/sdpb/build/src/sdp2input/write_output/bilinear_ba
sis/bilinear_form/bilinear_form.cxx.4.o']
../src/sdpb/SDP_Solver_Parameters/SDP_Solver_Parameters.cxx: In constructor ‘SDP_Solver_Parameters::SDP_Solver_Parameters(int, char**)’:
../src/sdpb/SDP_Solver_Parameters/SDP_Solver_Parameters.cxx:264:15: error: ‘ofstream’ is not a member of ‘boost::filesystem’
               boost::filesystem::ofstream ofs(out_directory / "out.txt");
               ^
```

which doesn't make any sense to me, since we have the Ubuntu libboost-filesystem-dev package installed, version 1_58_0, and that definitely *does* have ofstream in the boost::filesystem namespace.  I added to 

src/sdpb/SDP_Solver_Parameters/SDP_Solver_Parameters.cxx

#include <boost/filesystem/fstream.hpp>

and that compiles.

Then I get:

```
[89/94] Compiling src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx
12:18:40 runner ['mpicxx', '-Wall', '-Wextra', '-O3', '-D SDPB_VERSION_STRING="2.1.2-2-g7cadb27"', '-std=c++14', '-I/cm/shared/apps/elemental/spdb-fork/include', '-I/usr/include/libxml2', '-DHAVE_GMPXX_H=
1', '-DHAVE_MPFR_H=1', '-DHAVE_LIBXML2=1', '../src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx', '-c', '-o/home/dlang/spdb-build/sdpb/build/src/sdp2input/write_output/bilinear_ba
sis/bilinear_form/bilinear_form.cxx.4.o']
../src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx: In function ‘Boost_Float bilinear_form(const Damped_Rational&, const std::vector<boost::multiprecision::number<boost::multipre
cision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> > >&, const std::vector<std::pair<__gnu_cxx::__normal_iterator<const boost::multiprecision::number<boost::multipre
cision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> >*, std::vector<boost::multiprecision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::mult
iprecision::mpfr_allocation_type)1u> > > >, __gnu_cxx::__normal_iterator<const boost::multiprecision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation
_type)1u> >*, std::vector<boost::multiprecision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> > > > > >&, const std::vector<long int>&, c
onst std::vector<boost::multiprecision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> > >&, const std::vector<std::vector<boost::multiprec
ision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> > > >&, const int64_t&)’:
../src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx:53:59: error: no matching function for call to ‘boost::math::tools::polynomial<boost::multiprecision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> > >::polynomial(<brace-enclosed initializer list>)’
     pow(boost::math::tools::polynomial<Boost_Float>({0, 1}), m)),
                                                           ^
In file included from ../src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx:6:0:
/usr/include/boost/math/tools/polynomial.hpp:127:4: note: candidate: template<class U> boost::math::tools::polynomial<T>::polynomial(const boost::math::tools::polynomial<U>&)
    polynomial(const polynomial<U>& p)
    ^
/usr/include/boost/math/tools/polynomial.hpp:127:4: note:   template argument deduction/substitution failed:
../src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx:53:59: note:   couldn't deduce template parameter ‘U’
     pow(boost::math::tools::polynomial<Boost_Float>({0, 1}), m)),
                                                           ^
In file included from ../src/sdp2input/write_output/bilinear_basis/bilinear_form/bilinear_form.cxx:6:0:
/usr/include/boost/math/tools/polynomial.hpp:123:4: note: candidate: boost::math::tools::polynomial<T>::polynomial(const boost::math::tools::polynomial<T>&) [with T = boost::multiprecision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> >]
    polynomial(const polynomial& p)
    ^
/usr/include/boost/math/tools/polynomial.hpp:123:4: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘const boost::math::tools::polynomial<boost::multiprecision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> > >&’
/usr/include/boost/math/tools/polynomial.hpp:117:4: note: candidate: template<class U> boost::math::tools::polynomial<T>::polynomial(const U&)
    polynomial(const U& point)
    ^
[....etc.....]
```

Is there a minimum Boost version that we're missing?  Or can you provide any hints?

**Comment by wlandry** [comment]
I have updated Install.md to use the main branch.  I also added your include fix to 

    src/sdpb/SDP_Solver_Parameters/SDP_Solver_Parameters.cxx

Thanks.

At the Perimeter Institute, Erich Schnetter has already installed SDPB using Singularity.  Would that work for you?

In any event, for the problem you are running into now, boost might be too old.  I know that 1.55 is too old because it does not have pow(polynomial,int).  I use 1.67.  I had problems with 1.69.

Also, I think you will have ABI problems with the builtin boost.  The default compiler on Ubuntu 16.04 is 5.3.1, which defaults to C++98.  So your boost will be built with C++98.  The rest of SDPB needs to be built with C++14, and the ABI is different.  To build, I use commands like

    ./bootstrap.sh --prefix=$HOME/install
    ./b2 --prefix=$HOME/install -j 16 --reconfigure cxxstd=14
    ./b2 --prefix=$HOME/install -j 16  cxxstd=14 install
